astropy/coordinates/spectral_coordinate.py

```diff
@@ -0,0 +1,842 @@
+        # calculations that lack numerical precision.
+        observer_icrs = observer.transform_to(ICRS)
+        target_icrs = target.transform_to(ICRS)
+
+        pos_hat = SpectralCoord._norm_d_pos(observer_icrs, target_icrs)
+
+        d_vel = target_icrs.velocity - observer_icrs.velocity
+
+        vel_mag = np.dot(d_vel.d_xyz, pos_hat.xyz)
+
+        if NUMPY_LT_1_17:
+            vel_mag *= d_vel.d_xyz.unit
+
+        if as_scalar:
+            return vel_mag
+        else:
+            return vel_mag * pos_hat.xyz
+
+    @staticmethod
+    def _norm_d_pos(observer, target):
+        """
+        Calculate the normalized position vector between two frames.
+
+        Parameters
+        ----------
+        observer : `~astropy.coordinates.BaseCoordinateFrame` or `~astropy.coordinates.SkyCoord`
+            The observation frame or coordinate.
+        target : `~astropy.coordinates.BaseCoordinateFrame` or `~astropy.coordinates.SkyCoord`
+            The target frame or coordinate.
+
+        Returns
+        -------
+        pos_hat : `BaseRepresentation`
+            Position representation.
+        """
+        d_pos = (target.data.without_differentials() -
+                 observer.data.without_differentials()).to_cartesian()
+
+        dp_norm = d_pos.norm()
+
+        # Reset any that are 0 to 1 to avoid nans from 0/0
+        dp_norm.ravel()[dp_norm.ravel() == 0] = 1 * dp_norm.unit
+
+        pos_hat = d_pos / dp_norm
+
+        return pos_hat
+
+    def _change_observer_to(self, observer, target=None):
+        """
+        Moves the observer to the provided coordinate/frame.
+
+        Parameters
+        ----------
+        observer : `~astropy.coordinates.BaseCoordinateFrame` or `~astropy.coordinates.SkyCoord`
+            The new observation frame or coordinate.
+        target : `~astropy.coordinates.SkyCoord`, optional
+            The `~astropy.coordinates.SkyCoord` object representing the target of the observation.
+            If none given, defaults to currently defined target.
+
+        Returns
+        -------
+        new_coord : `SpectralCoord`
+            The new coordinate object representing the spectral data
+            transformed to the new observer frame.
+        """
+
+        if self.observer is None:
+            raise ValueError("No observer has been set, cannot change "
+                             "observer.")
+
+        target = self.target if target is None else target
+
+        # Check velocities and distance values on the new frame. This is
+        #  handled in the frame validation.
+        observer = self._validate_coordinate(observer)
+
+        # Calculate the initial and final los velocity
+        init_obs_vel = self._calculate_radial_velocity(self.observer, target)
+        fin_obs_vel = self._calculate_radial_velocity(observer, target)
+
+        line_of_sight_unit_vec = self._norm_d_pos(observer, target).xyz
```

Comment:
I'd really hope it would not be necessary to take the quantity here, i.e., do without `.xyz`. I think with my suggested edit to `_calculate_radial_velocity` it may not be necessary.

Reply:
Done